File: packages/pdf-cli/tiwater_pdf/runtime_identity.py
"""Non-mutating runtime capability and PDF identity evidence probes."""

from __future__ import annotations

import hashlib
from pathlib import Path
import re
from typing import Any

import fitz

from . import __version__
from .runtime_contract import identify_canonical_json_artifact, normalize_evidence


CONTRACT_VERSION = "1.0.0"
CAPABILITIES_SCHEMA_ID = (
    "https://tiwater.dev/contracts/runtime/runtime-capabilities.schema.json"
)
EVIDENCE_SCHEMA_ID = (
    "https://tiwater.dev/contracts/runtime/runtime-evidence-envelope.schema.json"
)
IDENTIFY_PAYLOAD_SCHEMA_ID = "tiwater.runtime.identify-payload"
PACKAGE_NAME = "tiwater-pdf"
RUNTIME_NAME = "tiwater-pdf"
SIGNATURE_KIND = "pdf-header-pymupdf-open"
PDF_MEDIA_TYPE = "application/pdf"
_PDF_HEADER = re.compile(rb"\A%PDF-(?P<version>(?:1\.[0-7]|2\.0))(?:\r\n|\r|\n)")


def capabilities() -> dict[str, Any]:
    """Return the versioned, non-mutating PDF runtime descriptor."""

    return {
        "schemaVersion": CONTRACT_VERSION,
        "descriptorType": "runtime-capabilities",
        "package": _package_identity(),
        "runtime": _runtime_identity(),
        "evidenceSchema": _schema_identity(EVIDENCE_SCHEMA_ID),
        "descriptorCommand": {
            "command": "capabilities",
            "arguments": ["--json"],
            "mutates": False,
        },
        "identifyProbe": {
            "command": "identify",
            "arguments": ["<input>", "--json"],
            "mutates": False,
            "outcomes": ["supported", "unsupported", "failed"],
        },
        "supportedKinds": [
            {
                "fileKind": "pdf",
                "mediaTypes": [PDF_MEDIA_TYPE],
                "signatureKinds": [SIGNATURE_KIND],
            }
        ],
        "commands": [
            {
                "name": "capabilities",
                "mutates": False,
                "outputSchema": _schema_identity(CAPABILITIES_SCHEMA_ID),
            },
            {
                "name": "identify",
                "mutates": False,
                "outputSchema": _schema_identity(EVIDENCE_SCHEMA_ID),
            },
            {
                "name": "extract-evidence",
                "mutates": False,
                "outputSchema": _schema_identity(EVIDENCE_SCHEMA_ID),
            },
        ],
        "identityPolicy": {
            "nativeIds": "runtime-native-only",
            "derivedIds": "deterministic-and-explicit",
            "containment": "parent-object-id-required-for-non-root",
        },
    }


def identify(file_path: str | Path) -> dict[str, Any]:
    """Identify PDF bytes without OCR, network access, or source mutation."""

    try:
        resolved_path = Path(file_path).expanduser().resolve(strict=True)
        source_bytes = resolved_path.read_bytes()
    except (OSError, RuntimeError, ValueError):
        return _evidence_envelope(
            status="failed",
            failure_stage="source-read",
            source=None,
            file_evidence=_file_evidence(
                signature_status="not-checked",
                signature_evidence=[],
            ),
            payload={"failureClass": "source-read-error"},
            errors=[
                {
                    "code": "source-read-failed",
                    "message": "The source bytes could not be read.",
                }
            ],
        )

    source = _source_identity(resolved_path, source_bytes)
    header = _PDF_HEADER.match(source_bytes)
    if header is None:
        return _unsupported(
            source,
            reason="pdf-header-mismatch",
            evidence=["pdf-header:mismatched"],
        )

    version = header.group("version").decode("ascii")
    try:
        with fitz.open(stream=source_bytes, filetype="pdf") as document:
            if not document.is_pdf:
                return _unsupported(
                    source,
                    reason="pymupdf-not-pdf",
                    evidence=[f"pdf-header:version={version}", "pymupdf:not-pdf"],
                )
            if document.is_repaired:
                return _unsupported(
                    source,
                    reason="pymupdf-repair-required",
                    evidence=[
                        f"pdf-header:version={version}",
                        "pymupdf:pdf-repair-required",
                    ],
                )
            encrypted = bool(document.needs_pass or document.is_encrypted)
    except (fitz.FileDataError, RuntimeError, TypeError, ValueError):
        return _unsupported(
            source,
            reason="pymupdf-open-rejected",
            evidence=[f"pdf-header:version={version}", "pymupdf:pdf-rejected"],
        )

    return _evidence_envelope(
        status="supported",
        failure_stage=None,
        source=source,
        file_evidence=_file_evidence(
            file_kind="pdf",
            media_type=PDF_MEDIA_TYPE,
            signature_status="matched",
            signature_evidence=[
                f"pdf-header:version={version}",
                "pymupdf:pdf-opened",
                f"pdf:encrypted={str(encrypted).lower()}",
            ],
        ),
        payload={"recognized": True, "fileKind": "pdf", "encrypted": encrypted},
        errors=[],
    )


def extraction_evidence(
    identify_evidence: dict[str, Any],
    report: dict[str, Any] | None,
) -> dict[str, Any]:
    """Bind normalized extraction nodes to a previously computed identity probe."""

    if identify_evidence["status"] == "supported":
        normalized = normalize_evidence(report or {})
        payload = normalized["payload"]
        objects = normalized["objects"]
    else:
        payload = {
            "identifyStatus": identify_evidence["status"],
            "reason": "source-not-supported-for-extraction",
            "nodes": [],
        }
        objects = []

    return _evidence_envelope(
        status=identify_evidence["status"],
        failure_stage=identify_evidence["failureStage"],
        source=identify_evidence["source"],
        file_evidence=identify_evidence["file"],
        payload=payload,
        errors=identify_evidence["errors"],
        probe="extract-evidence",
        objects=objects,
        payload_schema_id="tiwater.runtime.normalized-evidence",
    )


def _unsupported(
    source: dict[str, Any],
    *,
    reason: str,
    evidence: list[str],
) -> dict[str, Any]:
    return _evidence_envelope(
        status="unsupported",
        failure_stage=None,
        source=source,
        file_evidence=_file_evidence(
            signature_status="mismatched",
            signature_evidence=evidence,
        ),
        payload={"recognized": False, "reason": reason},
        errors=[],
    )


def _evidence_envelope(
    *,
    status: str,
    failure_stage: str | None,
    source: dict[str, Any] | None,
    file_evidence: dict[str, Any],
    payload: dict[str, Any],
    errors: list[dict[str, Any]],
    probe: str = "identify",
    objects: list[dict[str, Any]] | None = None,
    payload_schema_id: str = IDENTIFY_PAYLOAD_SCHEMA_ID,
) -> dict[str, Any]:
    return {
        "schemaVersion": CONTRACT_VERSION,
        "envelopeType": "runtime-evidence",
        "probe": probe,
        "status": status,
        "failureStage": failure_stage,
        "package": _package_identity(),
        "runtime": _runtime_identity(),
        "evidenceSchema": _schema_identity(EVIDENCE_SCHEMA_ID),
        "source": source,
        "file": file_evidence,
        "artifact": identify_canonical_json_artifact(
            payload,
            schema_id=payload_schema_id,
            schema_version=CONTRACT_VERSION,
        ),
        "payload": payload,
        "objects": objects or [],
        "warnings": [],
        "errors": errors,
    }


def _source_identity(resolved_path: Path, source_bytes: bytes) -> dict[str, Any]:
    sha256 = hashlib.sha256(source_bytes).hexdigest()
    return {
        "path": str(resolved_path),
        "sizeBytes": len(source_bytes),
        "sha256": sha256,
        "contentId": f"sha256:{sha256}",
    }


def _file_evidence(
    *,
    signature_status: str,
    signature_evidence: list[str],
    file_kind: str | None = None,
    media_type: str | None = None,
) -> dict[str, Any]:
    return {
        "fileKind": file_kind,
        "mediaType": media_type,
        "signature": {
            "status": signature_status,
            "kind": SIGNATURE_KIND,
            "evidence": signature_evidence,
        },
    }


def _package_identity() -> dict[str, str]:
    return {"name": PACKAGE_NAME, "version": __version__}


def _runtime_identity() -> dict[str, str]:
    return {"family": "pdf", "name": RUNTIME_NAME, "version": __version__}


def _schema_identity(schema_id: str) -> dict[str, str]:
    return {"id": schema_id, "version": CONTRACT_VERSION}

File: packages/xlsx-cli/RuntimeIdentity.cs
using System.IO.Compression;
using System.Reflection;
using System.Text.Json;
using System.Xml;
using System.Xml.Linq;
using DocumentFormat.OpenXml.Packaging;
using NPOI.HSSF.UserModel;
using Tiwater.RuntimeContracts;

namespace Dockit.Xlsx.Cli;

public static class XlsxRuntimeIdentity
{
    private const string PackageName = "tiwater.xlsx.cli";
    private const string RuntimeName = "tiwater-xlsx";
    private const string EvidenceSchemaId = "https://tiwater.dev/contracts/runtime/runtime-evidence-envelope.schema.json";
    private const string CapabilitiesSchemaId = "https://tiwater.dev/contracts/runtime/runtime-capabilities.schema.json";
    private const string PayloadSchemaId = "tiwater.runtime.identify-payload";
    private const string XlsxMediaType = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
    private const string XlsMediaType = "application/vnd.ms-excel";
    private const string SpreadsheetMainPartContentType =
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml";
    private const string XlsxSignatureKind = "ooxml-spreadsheet-main-part";
    private const string XlsSignatureKind = "ole-compound-hssf-workbook";
    private const string GenericSignatureKind = "spreadsheet-package-signature";

    private static readonly byte[] ZipLocalFileMagic = [0x50, 0x4b, 0x03, 0x04];
    private static readonly byte[] OleCompoundMagic = [0xd0, 0xcf, 0x11, 0xe0, 0xa1, 0xb1, 0x1a, 0xe1];
    private static readonly string PackageVersion = ResolvePackageVersion();
    private static readonly SchemaIdentity EvidenceSchema =
        new(EvidenceSchemaId, RuntimeContractVersions.EvidenceEnvelope);
    private static readonly SchemaIdentity PayloadSchema =
        new(PayloadSchemaId, RuntimeContractVersions.EvidenceEnvelope);

    public static RuntimeCapabilityDescriptor Capabilities() => new(
        RuntimeContractVersions.Capabilities,
        "runtime-capabilities",
        PackageIdentity(),
        RuntimeIdentity(),
        EvidenceSchema,
        new DiscoveryCommand("capabilities", ["--json"], false),
        new IdentifyProbe("identify", ["<input>", "--json"], false, ["supported", "unsupported", "failed"]),
        [
            new SupportedKind("xlsx", [XlsxMediaType], [XlsxSignatureKind]),
            new SupportedKind("xls", [XlsMediaType], [XlsSignatureKind]),
        ],
        [
            new RuntimeCommand(
                "capabilities",
                false,
                new SchemaIdentity(CapabilitiesSchemaId, RuntimeContractVersions.Capabilities)),
            new RuntimeCommand("identify", false, EvidenceSchema),
        ],
        new IdentityPolicy(
            "runtime-native-only",
            "deterministic-and-explicit",
            "parent-object-id-required-for-non-root"));

    public static RuntimeEvidenceEnvelope Identify(string filePath)
    {
        string resolvedPath;
        byte[] sourceBytes;
        try
        {
            resolvedPath = Path.GetFullPath(filePath);
            sourceBytes = File.ReadAllBytes(resolvedPath);
        }
        catch (Exception exception) when (
            exception is IOException
                or UnauthorizedAccessException
                or ArgumentException
                or NotSupportedException)
        {
            return CreateEnvelope(
                status: "failed",
                failureStage: "source-read",
                source: null,
                file: new RuntimeFileEvidence(
                    null,
                    null,
                    new SignatureEvidence("not-checked", GenericSignatureKind, [])),
                payload: JsonSerializer.SerializeToElement(
                    new { failureClass = "source-read-error" },
                    RuntimeJson.Options),
                errors: [new ContractFinding("source-read-failed", "The source bytes could not be read.")]);
        }

        var source = FileIdentity.IdentifyBytes(resolvedPath, sourceBytes);
        SignatureInspection inspection;
        try
        {
            inspection = InspectSignature(sourceBytes);
        }
        catch (Exception exception) when (
            exception is IOException
                or InvalidDataException
                or XmlException
                or OpenXmlPackageException)
        {
            return CreateEnvelope(
                status: "failed",
                failureStage: "signature-inspection",
                source,
                file: new RuntimeFileEvidence(
                    null,
                    null,
                    new SignatureEvidence("unknown", GenericSignatureKind, [])),
                payload: JsonSerializer.SerializeToElement(
                    new { failureClass = "signature-inspection-error" },
                    RuntimeJson.Options),
                errors: [new ContractFinding(
                    "signature-inspection-failed",
                    "The spreadsheet signature could not be inspected.")]);
        }

        if (!inspection.Matched)
        {
            return CreateEnvelope(
                status: "unsupported",
                failureStage: null,
                source,
                file: new RuntimeFileEvidence(
                    null,
                    null,
                    new SignatureEvidence("mismatched", inspection.SignatureKind, inspection.Evidence)),
                payload: JsonSerializer.SerializeToElement(
                    new { recognized = false, reason = inspection.Reason },
                    RuntimeJson.Options),
                errors: []);
        }

        return CreateEnvelope(
            status: "supported",
            failureStage: null,
            source,
            file: new RuntimeFileEvidence(
                inspection.FileKind,
                inspection.MediaType,
                new SignatureEvidence("matched", inspection.SignatureKind, inspection.Evidence)),
            payload: JsonSerializer.SerializeToElement(
                new { recognized = true, fileKind = inspection.FileKind },
                RuntimeJson.Options),
            errors: []);
    }

    private static SignatureInspection InspectSignature(ReadOnlyMemory<byte> sourceBytes)
    {
        if (StartsWith(sourceBytes.Span, ZipLocalFileMagic))
        {
            return InspectOpenXmlSpreadsheet(sourceBytes);
        }
        if (StartsWith(sourceBytes.Span, OleCompoundMagic))
        {
            return InspectLegacySpreadsheet(sourceBytes);
        }
        return SignatureInspection.Unsupported(
            GenericSignatureKind,
            "unrecognized-container-signature",
            ["container-signature:unrecognized"]);
    }

    private static SignatureInspection InspectOpenXmlSpreadsheet(ReadOnlyMemory<byte> sourceBytes)
    {
        var evidence = new List<string> { "zip:local-file-header" };
        try
        {
            using var archiveStream = new MemoryStream(sourceBytes.ToArray(), writable: false);
            using var archive = new ZipArchive(archiveStream, ZipArchiveMode.Read, leaveOpen: false);
            var contentTypesEntries = archive.Entries
                .Where(entry => entry.FullName == "[Content_Types].xml")
                .ToArray();
            if (contentTypesEntries.Length != 1)
            {
                evidence.Add($"[Content_Types].xml:count={contentTypesEntries.Length}");
                return SignatureInspection.Unsupported(
                    XlsxSignatureKind,
                    "content-types-part-missing-or-ambiguous",
                    evidence);
            }

            XDocument contentTypes;
            using (var contentTypesStream = contentTypesEntries[0].Open())
            using (var reader = XmlReader.Create(contentTypesStream, new XmlReaderSettings
            {
                DtdProcessing = DtdProcessing.Prohibit,
                XmlResolver = null,
            }))
            {
                contentTypes = XDocument.Load(reader, LoadOptions.None);
            }

            XNamespace contentTypesNamespace = "http://schemas.openxmlformats.org/package/2006/content-types";
            string mainPartName;
            using (var packageStream = new MemoryStream(sourceBytes.ToArray(), writable: false))
            using (var spreadsheet = SpreadsheetDocument.Open(packageStream, false))
            {
                if (spreadsheet.WorkbookPart is null)
                {
                    evidence.Add("spreadsheet-main-part:openxml-workbook-part-missing");
                    return SignatureInspection.Unsupported(
                        XlsxSignatureKind,
                        "spreadsheet-main-part-not-openable",
                        evidence);
                }
                mainPartName = spreadsheet.WorkbookPart.Uri.ToString();
            }

            var overrideContentTypes = contentTypes
                .Descendants(contentTypesNamespace + "Override")
                .Where(element => string.Equals(
                    (string?)element.Attribute("PartName"),
                    mainPartName,
                    StringComparison.Ordinal))
                .Select(element => (string?)element.Attribute("ContentType"))
                .ToArray();
            if (overrideContentTypes.Length > 1)
            {
                evidence.Add($"spreadsheet-main-part:override-count={overrideContentTypes.Length}");
                return SignatureInspection.Unsupported(
                    XlsxSignatureKind,
                    "spreadsheet-main-part-content-type-ambiguous",
                    evidence);
            }

            var effectiveContentType = overrideContentTypes.SingleOrDefault();
            if (effectiveContentType is null)
            {
                var extension = Path.GetExtension(mainPartName).TrimStart('.');
                var defaultContentTypes = contentTypes
                    .Descendants(contentTypesNamespace + "Default")
                    .Where(element => string.Equals(
                        (string?)element.Attribute("Extension"),
                        extension,
                        StringComparison.OrdinalIgnoreCase))
                    .Select(element => (string?)element.Attribute("ContentType"))
                    .ToArray();
                if (defaultContentTypes.Length == 1) effectiveContentType = defaultContentTypes[0];
            }
            if (!string.Equals(effectiveContentType, SpreadsheetMainPartContentType, StringComparison.Ordinal))
            {
                evidence.Add("spreadsheet-main-part:content-type-mismatch");
                return SignatureInspection.Unsupported(
                    XlsxSignatureKind,
                    "spreadsheet-main-part-content-type-mismatch",
                    evidence);
            }

            var archiveEntryName = mainPartName.TrimStart('/');
            var mainPartCount = archive.Entries.Count(entry => entry.FullName == archiveEntryName);
            if (mainPartCount != 1)
            {
                evidence.Add($"spreadsheet-main-part:entry-count={mainPartCount}");
                return SignatureInspection.Unsupported(
                    XlsxSignatureKind,
                    "spreadsheet-main-part-missing-or-ambiguous",
                    evidence);
            }

            evidence.Add($"[Content_Types].xml:{mainPartName}={SpreadsheetMainPartContentType}");
            evidence.Add($"part:{archiveEntryName}");
            evidence.Add("openxml:workbook-part-opened");
            return SignatureInspection.Supported(
                "xlsx",
                XlsxMediaType,
                XlsxSignatureKind,
                evidence);
        }
        catch (Exception exception) when (
            exception is InvalidDataException
                or XmlException
                or OpenXmlPackageException
                or IOException)
        {
            evidence.Add("spreadsheet-main-part:package-rejected");
            return SignatureInspection.Unsupported(
                XlsxSignatureKind,
                "spreadsheet-package-invalid",
                evidence);
        }
    }

    private static SignatureInspection InspectLegacySpreadsheet(ReadOnlyMemory<byte> sourceBytes)
    {
        var evidence = new List<string> { "ole-compound:magic=d0cf11e0a1b11ae1" };
        try
        {
            using var stream = new MemoryStream(sourceBytes.ToArray(), writable: false);
            using var workbook = new HSSFWorkbook(stream);
            _ = workbook.NumberOfSheets;
            evidence.Add("npoi-hssf:workbook-opened");
            return SignatureInspection.Supported("xls", XlsMediaType, XlsSignatureKind, evidence);
        }
        catch (Exception exception) when (
            exception is IOException
                or InvalidDataException
                or ArgumentException
                or InvalidOperationException
                or IndexOutOfRangeException
                or NPOI.Util.RecordFormatException)
        {
            evidence.Add("npoi-hssf:workbook-rejected");
            return SignatureInspection.Unsupported(
                XlsSignatureKind,
                "ole-compound-not-hssf-workbook",
                evidence);
        }
    }

    private static bool StartsWith(ReadOnlySpan<byte> bytes, ReadOnlySpan<byte> prefix) =>
        bytes.Length >= prefix.Length && bytes[..prefix.Length].SequenceEqual(prefix);

    private static RuntimeEvidenceEnvelope CreateEnvelope(
        string status,
        string? failureStage,
        FileContentIdentity? source,
        RuntimeFileEvidence file,
        JsonElement payload,
        IReadOnlyList<ContractFinding> errors)
    {
        var retainedPayload = payload.Clone();
        return new RuntimeEvidenceEnvelope(
            RuntimeContractVersions.EvidenceEnvelope,
            "runtime-evidence",
            "identify",
            status,
            failureStage,
            PackageIdentity(),
            RuntimeIdentity(),
            EvidenceSchema,
            source,
            file,
            EvidenceEnvelope.IdentifyCanonicalJson(retainedPayload, PayloadSchema),
            retainedPayload,
            [],
            [],
            errors);
    }

    private static PackageIdentity PackageIdentity() => new(PackageName, PackageVersion);

    private static RuntimeIdentity RuntimeIdentity() => new("office", RuntimeName, PackageVersion);

    private static string ResolvePackageVersion()
    {
        var informationalVersion = typeof(XlsxRuntimeIdentity).Assembly
            .GetCustomAttribute<AssemblyInformationalVersionAttribute>()?
            .InformationalVersion;
        if (string.IsNullOrWhiteSpace(informationalVersion))
        {
            throw new InvalidOperationException("XLSX package version metadata is unavailable.");
        }
        return informationalVersion.Split('+', 2)[0];
    }

    private sealed record SignatureInspection(
        bool Matched,
        string? FileKind,
        string? MediaType,
        string SignatureKind,
        string Reason,
        IReadOnlyList<string> Evidence)
    {
        public static SignatureInspection Supported(
            string fileKind,
            string mediaType,
            string signatureKind,
            IReadOnlyList<string> evidence) =>
            new(true, fileKind, mediaType, signatureKind, "matched", evidence);

        public static SignatureInspection Unsupported(
            string signatureKind,
            string reason,
            IReadOnlyList<string> evidence) =>
            new(false, null, null, signatureKind, reason, evidence);
    }
}
